java.lang.IndexOutOfBoundsException: Index: 1, Size: 0

- [ ] I have searched the [issues](https://github.com/seata/seata/issues) of this repository and believe that this is not a duplicate.

### Ⅰ. Issue Description
压测中出现,批量rm注册分支后,并行处理时数组越界
```
2023-02-01 20:33:18.344 ERROR --- [   ServerHandlerThread_1_6_500] i.s.c.r.p.s.ServerOnRequestProcessor     : handle request error: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
==>
java.util.concurrent.ExecutionException: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
	at java.base/java.util.concurrent.CompletableFuture.reportGet(CompletableFuture.java:396)
	at java.base/java.util.concurrent.CompletableFuture.get(CompletableFuture.java:2073)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.onRequestMessage(ServerOnRequestProcessor.java:201)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.process(ServerOnRequestProcessor.java:123)
	at io.seata.core.rpc.netty.AbstractNettyRemoting.lambda$processMessage$2(AbstractNettyRemoting.java:280)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
	at java.base/java.util.concurrent.CopyOnWriteArrayList.add(CopyOnWriteArrayList.java:451)
	at io.seata.core.rpc.processor.server.ServerOnRequestProcessor.lambda$onRequestMessage$1(ServerOnRequestProcessor.java:191)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1804)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1796)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:373)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1182)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1655)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1622)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:165)
<==

```

### Ⅱ. Describe what happened

  If there is an exception, please attach the exception trace:

```
Just paste your stack trace here!
```


### Ⅲ. Describe what you expected to happen


### Ⅳ. How to reproduce it (as minimally and precisely as possible)

1. xxx
2. xxx
3. xxx

### Ⅴ. Anything else we need to know?


### Ⅵ. Environment:

- JDK version :
- Seata version:  
- OS :
- Others:

--- a/changes/en-us/develop.md
+++ b/changes/en-us/develop.md
@@ -14,6 +14,7 @@ Add changes here for all PR submitted to the develop branch.
 - [[#5266](https://github.com/seata/seata/pull/5265)] fix server console has queried the released lock
 - [[#5245](https://github.com/seata/seata/pull/5245)] fix the incomplete dependency of distribution module
 - [[#5239](https://github.com/seata/seata/pull/5239)] fix `getConfig` throw `ClassCastException` when use JDK proxy
+- [[#5281](https://github.com/seata/seata/pull/5281)] parallel request handle throw IndexOutOfBoundsException
 
 ### optimize:
 - [[#5208](https://github.com/seata/seata/pull/5208)] optimize throwable getCause once more
@@ -44,6 +45,7 @@ Thanks to these contributors for their code commits. Please report an unintended
 - [pengten](https://github.com/pengten)
 - [wangliang181230](https://github.com/wangliang181230)
 - [GoodBoyCoder](https://github.com/GoodBoyCoder)
+- [a364176773](https://github.com/a364176773)
 
 
 Also, we receive many valuable issues, questions and advices from our community. Thanks for you all.

--- a/changes/zh-cn/develop.md
+++ b/changes/zh-cn/develop.md
@@ -14,6 +14,7 @@
 - [[#5266](https://github.com/seata/seata/pull/5265)] 修复控制台全局锁查询接口查到了已释放的锁
 - [[#5245](https://github.com/seata/seata/pull/5245)] 修复不完整的distribution模块依赖
 - [[#5239](https://github.com/seata/seata/pull/5239)] 修复当使用JDK代理时，`getConfig` 方法获取部分配置时抛出 `ClassCastException` 异常的问题
+- [[#5281](https://github.com/seata/seata/pull/5281)] 修复并行rm请求处理时数组索引越界问题
 
 ### optimize:
 - [[#5208](https://github.com/seata/seata/pull/5208)] 优化多次重复获取Throwable#getCause问题
@@ -44,6 +45,7 @@
 - [pengten](https://github.com/pengten)
 - [wangliang181230](https://github.com/wangliang181230)
 - [GoodBoyCoder](https://github.com/GoodBoyCoder)
+- [a364176773](https://github.com/a364176773)
 
 
 同时，我们收到了社区反馈的很多有价值的issue和建议，非常感谢大家。

--- a/core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java
+++ b/core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java
@@ -179,18 +179,16 @@ public class ServerOnRequestProcessor implements RemotingProcessor, Disposable {
                     }
                 }
             } else {
-                List<AbstractResultMessage> results = new CopyOnWriteArrayList<>();
-                List<CompletableFuture<Void>> completableFutures = null;
+                List<AbstractResultMessage> results = new ArrayList<>();
+                List<CompletableFuture<AbstractResultMessage>> completableFutures = null;
                 for (int i = 0; i < ((MergedWarpMessage)message).msgs.size(); i++) {
                     if (PARALLEL_REQUEST_HANDLE) {
                         if (completableFutures == null) {
                             completableFutures = new ArrayList<>();
                         }
                         int finalI = i;
-                        completableFutures.add(CompletableFuture.runAsync(() -> {
-                            results.add(finalI, handleRequestsByMergedWarpMessage(
-                                ((MergedWarpMessage)message).msgs.get(finalI), rpcContext));
-                        }));
+                        completableFutures.add(CompletableFuture.supplyAsync(() -> handleRequestsByMergedWarpMessage(
+                            ((MergedWarpMessage)message).msgs.get(finalI), rpcContext)));
                     } else {
                         results.add(i,
                             handleRequestsByMergedWarpMessage(((MergedWarpMessage)message).msgs.get(i), rpcContext));
@@ -198,7 +196,9 @@ public class ServerOnRequestProcessor implements RemotingProcessor, Disposable {
                 }
                 if (CollectionUtils.isNotEmpty(completableFutures)) {
                     try {
-                        CompletableFuture.allOf(completableFutures.toArray(new CompletableFuture[0])).get();
+                        for (CompletableFuture<AbstractResultMessage> completableFuture : completableFutures) {
+                            results.add(completableFuture.get());
+                        }
                     } catch (InterruptedException | ExecutionException e) {
                         LOGGER.error("handle request error: {}", e.getMessage(), e);
                     }

--- a/core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java
+++ b/core/src/main/java/io/seata/core/rpc/processor/server/ServerOnRequestProcessor.java
@@ -24,7 +24,6 @@ import java.util.concurrent.BlockingQueue;
 import java.util.concurrent.CompletableFuture;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
-import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.ExecutionException;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.LinkedBlockingQueue;

